vnext/JSI/Shared/ChakraApi.h

```diff
@@ -0,0 +1,468 @@
+
+    JsRefHolder(JsRefHolder const &other) noexcept;
+    JsRefHolder(JsRefHolder &&other) noexcept;
+
+    JsRefHolder &operator=(JsRefHolder const &other) noexcept;
+    JsRefHolder &operator=(JsRefHolder &&other) noexcept;
+
+    ~JsRefHolder() noexcept;
+
+    operator JsRef() const noexcept {
+      return m_ref;
+    }
+
+   private:
+    JsRef m_ref{JS_INVALID_REFERENCE};
+  };
+
+  /**
+   * @brief Interface that helps override the exception being thrown.
+   *
+   * The ChakraApi uses the ExceptionThrowerHolder to retrieve the ExceptionThrower
+   * instance from the current thread.
+   */
+  struct IExceptionThrower {
+    /**
+     * @brief Throws an exception based on the errorCore.
+     *
+     * The base implementation throws the generic std::exception.
+     * The method can be overridden in derived class to provide more specific exceptions.
+     */
+    [[noreturn]] virtual void ThrowJsExceptionOverride(JsErrorCode errorCode, JsValueRef jsError) = 0;
+
+    /**
+     * @brief Throws an exception with the provided errorMessage.
+     *
+     * The base implementation throws the generic std::exception.
+     * The method can be overridden in derived class to provide more specific exceptions.
+     */
+    [[noreturn]] virtual void ThrowNativeExceptionOverride(char const *errorMessage) = 0;
+  };
+
+  /**
+   * @brief A RAII class to hold IExceptionThrower instance in the thread local variable.
+   */
+  struct ExceptionThrowerHolder {
+    ExceptionThrowerHolder(IExceptionThrower *exceptionThrower) noexcept
+        : m_previous{std::exchange(tls_exceptionThrower, exceptionThrower)} {}
+
+    ~ExceptionThrowerHolder() noexcept {
+      tls_exceptionThrower = m_previous;
+    }
+
+    static IExceptionThrower *Get() noexcept {
+      return tls_exceptionThrower;
+    }
+
+   private:
+    static thread_local IExceptionThrower *tls_exceptionThrower;
+    IExceptionThrower *m_previous;
+  };
+
+  /**
+   * @brief Throws JavaScript exception with provided errorCode.
+   */
+  static void ThrowJsException(JsErrorCode errorCode);
+
+  /**
+   * @brief Throws native exception with provided message.
+   */
+  static void ThrowNativeException(char const *errorMessage);
+
+  /**
+   * @brief Adds a reference to a garbage collected object.
+   */
+  static uint32_t AddRef(JsRef ref);
+
+  /**
+   * @brief Releases a reference to a garbage collected object.
+   */
+  static uint32_t Release(JsRef ref);
+
+  /**
+   * @brief Creates a script context for running scripts.
+   */
+  static JsContextRef CreateContext(JsRuntimeHandle runtime);
+
+  /**
+   * @brief Gets the current script context on the thread..
+   */
+  JsContextRef GetCurrentContext();
+
+  /**
+   * @brief Sets the current script context on the thread.
+   */
+  void SetCurrentContext(JsContextRef context);
+
+  /**
+   * @brief Gets the property ID associated with the name.
+   */
+  static JsPropertyIdRef GetPropertyIdFromName(wchar_t const *name);
```

Comment:
>wchar_t const  [](start = 47, length = 14)

use PCWSTR to get the __null_terminated annotation?

Reply:
I prefer not to use macro for type names. We can change it later if need when we add annotations to all methods.

---
In reply to: [470460576](https://github.com/microsoft/react-native-windows/pull/5710#discussion_r470460576) [](ancestors = 470460576)